azurerm/resource_arm_dns_a_record.go

```diff
@@ -150,10 +169,15 @@ func resourceArmDnsARecordRead(d *schema.ResourceData, meta interface{}) error {
 	d.Set("zone_name", zoneName)
 	d.Set("ttl", resp.TTL)
 	d.Set("fqdn", resp.Fqdn)
+	d.Set("target_resource_id", (resp.TargetResource).ID)
 
-	if err := d.Set("records", flattenAzureRmDnsARecords(resp.ARecords)); err != nil {
-		return err
+	// Only flatten DNS records if they are present in the resource, e.g. not for alias records
+	if resp.ARecords != nil {
```

Comment:
If I don't cover this then the flattenAzureRmDnsARecords returns an error:

```panic: runtime error: invalid memory address or nil pointer dereference```

Again, if it is null, should I set ```records``` to be an empty slice?

Reply:
Ohhh, that looks like a bug in [flattenAzureRmDnsARecords](https://github.com/terraform-providers/terraform-provider-azurerm/blob/d415d3aaa6524f78dacaf4cca1de22e549f4de7f/azurerm/resource_arm_dns_a_record.go#L183). We should just be able to create that slice of strings with a length of 0 instead. 

We could also add another nil check [here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/d415d3aaa6524f78dacaf4cca1de22e549f4de7f/azurerm/resource_arm_dns_a_record.go#L187) and check that record isn't nil before pointing to `*record.Ipv4Address`.  

Apologies for all this extra nil checking. This is very important to prevent crashes like you saw.

Reply:
Yeah, I thought that. It actually panics [here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/d415d3aaa6524f78dacaf4cca1de22e549f4de7f/azurerm/resource_arm_dns_a_record.go#L183) on the creation of the results slice, as the pointer is to a null location. 

This is what I came up with to fix the issue:

```go
func flattenAzureRmDnsARecords(records *[]dns.ARecord) []string {
	if records != nil {
		results := make([]string, 0, len(*records))
		for _, record := range *records {
			results = append(results, *record.Ipv4Address)
		}
		return results
	}

	return make([]string, 0, 0)
}
```

Reply:
Looks good. Can we get away with defining results at the top of the function like `results := make([]string, 0, 0)` and then return that? Also we should nil check `record` before appending `*record.Ipv4Address` 

```
func flattenAzureRmDnsARecords(records *[]dns.ARecord) []string {
        results := make([]string, 0, 0)
	if records != nil {
		for _, record := range *records {
                         if record != nil {
			     results = append(results, *record.Ipv4Address)
                         }
		}
	}

	return results
}
```

Will that work the same?

Reply:
It will with ```record.Ipv4Address != nil```

Thanks again :)

Will modify and push later.